inc/az_ulib_ustream_forward_base.h

```diff
@@ -0,0 +1,374 @@
+ *          necessary for it to be passed to the ustream_forward.
+ */
+struct az_ulib_ustream_forward_tag
+{
+  /**
+   * @brief Control Block.
+   *
+   * The #az_ulib_ustream_forward_data_cb* on which this instance operates on.
+   */
+  az_ulib_ustream_forward_data_cb* control_block;
+
+  /* Instance controls. */
+
+  /** The #offset_t used as the logical position for this instance. */
+  offset_t offset_diff;
+
+  /** The #offset_t used to keep track of the current position (next returned position). */
+  offset_t inner_current_position;
+
+  /** The `size_t` with the length of the data in the control_block. */
+  size_t length;
+};
+
+/**
+ * @brief   Check if a handle is the same type of the API.
+ *
+ *  It will return true if the handle is valid and it is the same type of the API. It will
+ *      return false if the handle is `NULL` or not the correct type.
+ */
+#define AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(handle, type_api)                                             \
+  (!((handle == NULL) || (handle->control_block == NULL) || (handle->control_block->api == NULL) \
+     || (handle->control_block->api != &type_api)))
+
+/**
+ * @brief Copy data directly from source (provider) to destination (consumer)
+ * 
+ *  The `az_ulib_ustream_forward_flush` API will copy ALL of the contents of the data from the
+ *    source (provider) directly to the destination buffer provided by the consumer in the 
+ *    `#az_ulib_flush_callback* push_callback`.
+ *  
+ *  In the event that the source (provider) memory is not immediately available without a call
+ *    to an external API (e.g. an HTTP connection to a blob storage provider), the flush operation
+ *    will call this external API (e.g. http_response_body_get) in a loop, invoking the
+ *    `push_callback` each pass of the loop.
+ *  
+ *  The `az_ulib_ustream_forward_flush` API will provide the consumer a pointer to the memory to be
+ *    copied in the form of a `const az_span* const` buffer, to be consumed in the `push_callback`
+ *    function implemented by the consumer. 
+ * 
+ * @note: It is the consumer's responsibility to increment any write offset referenced in the
+ *        context inside the push_callback implementation.
+ * 
+ *  The `az_ulib_ustream_forward_flush` API shall meet the following minimum requirements:
+ *      - The flush operation create a `const az_span* const` from the `Data Source` and hand it
+ *        off to the `push_callback`
+ *      - If the totality of `Data Source` is not available in the provided memory location,
+ *        the API shall request the next portion of the data using the appropriate external API. 
+ *        This process shall be repeated in a loop until the totality of the `Data Source` has been
+ *        copied.
+ *      - If there is no more data to return, the flush shall return #AZ_ULIB_EOF and an empty
+ *        `const az_span* const` shall be created and passed to the `push_callback`.
```

Comment:
I don't think we should do that, `flush` will block the caller until all data was pushed to the push_callback. At the end of the data, `flush` will just return and will not call the push_callback anymore.

Reply:
What if there is a separate thread that wants to access the location being written to and we are looking to protect the memory until the callback is done writing?